lib/improver/tests/utilities/test_indexing_operations.py

```diff
@@ -0,0 +1,183 @@
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+"""Unit tests for the plugins and functions within indexing_operations.py"""
+
+import unittest
+import numpy as np
+
+from iris.tests import IrisTest
+from improver.utilities.indexing_operations import choose
+
+
+class Test_choose(IrisTest):
+
+    """Test the choose function behaves as expected, giving the same results
+    as the numpy choose method, but without the 32 leading dimensions limit."""
+
+    def setUp(self):
+        """Set up the data arrays."""
+        self.data = np.arange(132).reshape(33, 2, 2) + 1
+        self.small_data = np.arange(12).reshape(3, 2, 2) + 1
+
+    def test_single_index(self):
+        """Test that an array of indices containing only one value returns the
+        expected values."""
+        index_array = [0]
+        expected = np.array([[1, 2], [3, 4]])
+        result = choose(index_array, self.data)
+        self.assertArrayEqual(result, expected)
+        self.assertEqual(result.shape, expected.shape)
+
+    def test_1D_index_array_columns(self):
+        """Test that a 1D array of indices returns the expected values."""
+        index_array = [0, 1]
+        expected = np.array([[1, 6], [3, 8]])
+        result = choose(index_array, self.data)
+        self.assertArrayEqual(result, expected)
+        self.assertEqual(result.shape, expected.shape)
+
+    def test_2D_index_array_rows_only(self):
+        """Test that a 2D array of indices, arranged as a single column,
+        returns the expected values."""
+        index_array = [[0], [1]]
+        expected = np.array([[1, 2], [7, 8]])
+        result = choose(index_array, self.data)
+        self.assertArrayEqual(result, expected)
+        self.assertEqual(result.shape, expected.shape)
+
+    def test_2D_index_array(self):
+        """Test that a 2D array of indices returns the expected values."""
+        index_array = [[0, 1], [1, 2]]
+        expected = np.array([[1, 6], [7, 12]])
+        result = choose(index_array, self.data)
+        self.assertArrayEqual(result, expected)
+        self.assertEqual(result.shape, expected.shape)
+
+    def test_3D_index_array(self):
+        """Test that a 3D array of indices returns the expected values. Note
+        that in this case there is an additional dimension in the returned
+        array as the index_array has as many dimensions as the data array."""
+        index_array = [[[0, 1], [1, 2]]]
+        expected = np.array([[[1, 6], [7, 12]]])
+        result = choose(index_array, self.data)
+        self.assertArrayEqual(result, expected)
+        self.assertEqual(result.shape, expected.shape)
```

Comment:
I'd like to see a test for a varying 3D array of indices, where the index array matches the shape of the data array (using `self.small_data`), since this is the use we're expecting to make of this function in ECC.

Reply:
Done. This is the only functionality that works once I strip out the other features you do not want, so all the tests are now of this form.

Reply:
Note that the figure explaining the behaviour is no longer valid as the broadcasting has been removed, as such this figure will not be linked in the tests as it no longer matches the tests.

Reply:
I recreated the figure, and mention it in the doc string. We don't have sphinx-style documentation for the unit tests, so I have not embedded it in any other way.